rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,865 @@
+    1. Combine features using `RasaFeatureCombiningLayer`.
+    2. Apply a dense layer(s) to the combined features.
+    3. Optionally (during training for the `text` attribute), apply masking to the
+        features and create further helper variables for masked language modeling.
+    4. Embed the features using a transformer, effectively going from variable-length
+        sequences of features to fixed-size input example embeddings.
+
+    Arguments:
+        attribute: Name of attribute (e.g. `text` or `label`) whose features will be
+            processed.
+        attribute_signature: A dictionary containing two lists of `FeatureSignature`s,
+            one for each feature type of the given attribute.
+        config: A model config used for correctly parametrising the underlying layers.
+
+    Input shape:
+        Tuple of four tensor inputs:
+            sequence_features: List of 3-D dense or sparse tensors, each with shape
+                `(batch_size, max_seq_length, input_dim)` where `input_dim` can be
+                different for sparse vs dense tensors.
+            sentence_features: List of 3-D dense or sparse tensors, each with shape
+                `(batch_size, 1, input_dim)` where `input_dim` can be different for
+                sparse vs dense tensors, and can differ from that in `sequence_features`.
+            sequence_feature_lengths: Dense tensor of shape `(batch_size, )` containing
+                the real sequence lengths for the sequence-level features, i.e. the
+                numbers of real (not padding) tokens for each example in the batch.
+
+    Output shape:
+        outputs: `(batch_size, seq_length, units)` where `units` matches the underlying
+            transformer's output size if the transformer has some layers, otherwise
+            `units` matches that of the `Ffnn` block applied to the combined features, or
+            it's the output size of the underlying `RasaFeatureCombiningLayer` if the
+            `Ffnn` block has 0 layers. `seq_length` is the sum of the sequence dimension
+            sizes of sequence- and sentence-level features (for details, see the output
+            shape of `RasaFeatureCombiningLayer`). If both feature types are present,
+            then `seq_length` will be 1 + the length of the longest sequence of real
+            tokens across all examples in the given batch.
+        seq_sent_features: `(batch_size, seq_length, hidden_dim)`, where `hidden_dim` is
+            the output size of the underlying `Ffnn` block, or the output size of the
+            underlying `RasaFeatureCombiningLayer` if the `Ffnn` block has 0 layers.
+        mask_combined_sequence_sentence: `(batch_size, seq_length, hidden_dim)`
+        token_ids: `(batch_size, seq_length, id_dim)` where id_dim is unimportant and
+            it's the last-dimension size of the first sequence-level dense feature
+            if any is present, and 2 otherwise. Empty tensor if not doing MLM.
```

Comment:
From the description I don't get what you mean 🙈

Reply:
Yeah, I'll try to rephrase it 🙂

Reply:
@dakshvar22 how about this?
> `id_dim` is 2 when no dense sequence-level features are present. Otherwise, it's arbitrarily chosen to match the last dimension size of the first dense sequence-level feature in the input list of features.

Reply:
Sounds good 👍